flink-libraries/flink-state-processing-api/src/main/java/org/apache/flink/state/api/WritableSavepoint.java

```diff
@@ -90,38 +93,37 @@ public final void write(String path) {
 
 		List<OperatorState> existingOperators = metadata.getExistingOperators();
 
-		DataSet<OperatorState> finalOperatorStates = unionOperatorStates(newOperatorStates, existingOperators);
-
-		finalOperatorStates
-			.reduceGroup(new MergeOperatorStates(metadata.getMasterStates()))
-			.name("reduce(OperatorState)")
-			.output(new SavepointOutputFormat(savepointPath))
-			.name(path);
-	}
-
-	private DataSet<OperatorState> unionOperatorStates(DataSet<OperatorState> newOperatorStates, List<OperatorState> existingOperators) {
 		DataSet<OperatorState> finalOperatorStates;
 		if (existingOperators.isEmpty()) {
 			finalOperatorStates = newOperatorStates;
 		} else {
-			DataSet<OperatorState> wrappedCollection = newOperatorStates
-				.getExecutionEnvironment()
+			DataSet<OperatorState> existingOperatorStates = newOperatorStates.getExecutionEnvironment()
 				.fromCollection(existingOperators);
 
-			finalOperatorStates = newOperatorStates.union(wrappedCollection);
+			existingOperatorStates
+				.flatMap(new StatePathExtractor())
```

Comment:
We need to pin the parallelism of this to `1`. The collection source only runs a parallelism `1` so it will chain and then the paths will redistribute to the copy method. Otherwise we are not properly distributing this work. 

```suggestion
				.flatMap(new StatePathExtractor())
				.setParallelism(1)
```

Reply:
If we have a huge savepoint (e.g., in TB range) to process, do we also do with parallelism=1?  meaning only one thread does the file copy, potentially many of them.

Reply:
I misunderstood. It is for `.flatMap(new StatePathExtractor())`, not file copy.

Reply:
fixed